craft_archives/repo/apt_sources_manager.py

```diff
@@ -324,15 +324,45 @@ def install_package_repository_sources(
         raise RuntimeError(f"unhandled package repository: {package_repository!r}")
 
 
+def _is_deb822_default() -> bool:
+    """Return true if the default sources are in deb822 format."""
+    # TODO
+
+    return bool("You're using core24, right?")
+
+
+_COMPATIBLE_ARCHITECTURES = {
+    "amd64": ["amd64", "i386"],
+    "i386": ["amd64", "i386"],
+    "arm64": ["arm64", "armhf"],
+    "armhf": ["arm64", "armhf"],
+}
+
+
 def _add_architecture(architectures: List[str], root: Path) -> None:
-    """Add package repository architecture."""
+    """Add package repository architecture.
+
+    For systems that who default to deb822 sources, `dpkg --add-architecture`
+    works for any architecture. For systems that use the traditional sources.list,
+    only compatible architectures can be added.
+    """
     current_arch = _get_current_architecture()
-    compatible_pairs = {"amd64": "i386", "arm64": "armhf"}
+
+    # Sources in 'ubuntu.sources' don't list architectures, so apt assumes the default
+    # repository can serve all architectures. These need to be restricted before adding
+    # other architectures with `dpkg --add-architecture`.
+    if _is_deb822_default():
```

Comment:
For a non-experimental version, maybe we just replace `sources.list` with a deb822 format? dpkg supports it at least as far back as focal

Reply:
I think support for older releases is a good idea, whether we edit the old format or convert it to deb822.  

Most likely, we will get a request for core22 snaps.  When that happens, we can create an issue for it.